File: problems/p0242_valid_anagram/solution.cpp
#include <unordered_map>
#include <string>
#include <algorithm>
#include <vector>
#include <iostream>

class Solution {
public:
    bool isAnagram(std::string s, std::string t) {
        if (s.length() != t.length()) {
            return false;
        }
        std::unordered_map<char, int> char_count;
        // Since both strings are of equal length, we can iterate through both simultaneously
        for (unsigned int i = 0; i < s.length(); ++i) {
            ++char_count[s[i]];
            --char_count[t[i]];
        }
        for (const auto& [_, count] : char_count) {
            if (count != 0) {
                return false;
            }
        }
        return true;
    }
};

struct TestCase {
    std::string s;
    std::string t;
    bool expected;
};


int test() {
    Solution sol;
    TestCase testcases[] = {
        {"anagram", "nagaram", true},
        {"rat", "car", false},
        {"", "", true},
        {"a", "b", false},
        {"s", "s", true},
        {"a", "aaa", false}
    };

    int error_found = 0;

    for (const auto& test : testcases) {
        bool result = sol.isAnagram(test.s, test.t);
        if (result != test.expected) {
            // Use printf or println in stderr
            std::fprintf(stderr,
                "Test failed for input s: \"%s\", t: \"%s\". Expected %s but got %s.\n",
                test.s.c_str(), test.t.c_str(),
                test.expected ? "true" : "false",
                result ? "true" : "false");

            ++error_found;
        }
    }

    return error_found;
}

int main([[maybe_unused]] int argc, [[maybe_unused]] char* argv[]) {
    int errors = test();
    if (errors == 0) {
        std::cout << "All test cases passed!" << std::endl;
    } else {
        std::cerr << errors << " test case(s) failed." << std::endl;
        return 1;
    }
    return 0;
}

File: problems/p0242_valid_anagram/solution.rs
#![allow(unused)]

use std::collections::hash_map;
pub struct Solution;

impl Solution {
    pub fn is_anagram_sort(s: &str, t: &str) -> bool {
        let s_sort = {
            let mut chars: Vec<char> = s.chars().collect();
            chars.sort();
            chars
        };
        let t_sort = {
            let mut chars: Vec<char> = t.chars().collect();
            chars.sort();
            chars
        };
        s_sort == t_sort
    }

    pub fn is_anagram_linear(s: &str, t: &str) -> bool {
        let mut map_s = hash_map::HashMap::new();
        for c in s.chars() {
            *map_s.entry(c).or_insert(0) += 1;
        }
        let mut map_t = hash_map::HashMap::new();
        for c in t.chars() {
            *map_t.entry(c).or_insert(0) += 1;
        }
        map_s == map_t
    }
}

#[cfg(test)]
mod tests {
    struct TestCase {
        s: &'static str,
        t: &'static str,
        expected: bool,
    }
    use std::vec;

    use super::*;
    #[test]
    fn test_solution() {
        let cases = vec![
            TestCase {
                s: "anagram",
                t: "nagaram",
                expected: true,
            },
            TestCase {
                s: "rat",
                t: "car",
                expected: false,
            },
            TestCase {
                s: "",
                t: "",
                expected: true,
            },
            TestCase {
                s: "a",
                t: "b",
                expected: false,
            },
            TestCase {
                s: "s",
                t: "s",
                expected: true,
            },
            TestCase {
                s: "a",
                t: "aaa",
                expected: false,
            },
        ];
        for case in cases {
            println!("Testing s: {}, t: {}", case.s, case.t);
            assert_eq!(Solution::is_anagram_sort(case.s, case.t), case.expected);
            assert_eq!(Solution::is_anagram_linear(case.s, case.t), case.expected);
        }
    }
}
